forte/data/data_store.py

```diff
@@ -493,6 +498,137 @@ def get_entry(self, tid: int) -> Tuple[List, int, int]:
+
+        The precedence of those values indicates their priority in the min heap
+        ordering.
+        For example, if two entries have both the same begin and end field,
+        then their order is
+        decided by the order of user input type_name (the type that first
+        appears in the target type list will return first).
+        For entries that have the exact same `begin`, `end` and `type_name`,
+        the order will be determined arbitrarily.
+
+        Args:
+            type_names (List[str]): a list of string type names
+
+        Returns:
+
+            An iterator of entry elements.
+
+        """
+
+        n = len(type_names)
+        # suppose the length of type_names is N and the length of entry list of
+        # one type is M
+        # then the time complexity of using min-heap to iterate
+        # is O(M*log(N))
+
+        # Initialize the first entry of all entry lists
+        # it avoids empty entry lists or non-existant entry list
+        first_entries = []
+
+        for tn in type_names:
+            try:
+                first_entries.append(self.__elements[tn][0])
+            except KeyError as e:  # self.__elements[tn] will be catched here.
+                raise ValueError(
+                    f"Input argument `type_names` to the function contains"
+                    f" a type name [{tn}], which is not recognized."
+                    f" Please input available ones in this DataStore"
+                    f" object: {list(self.__elements.keys())}"
+                ) from e
+            except IndexError as e:  # self.__elements[tn][0] will be catched here.
+                raise ValueError(
+                    f"Entry list of type name, {tn} which is "
+                    " one list item of input argument `type_names`,"
+                    " is empty. Please check data in this DataStore). "
+                    " to see if empty lists are expected"
+                    f" or remove {tn} from input parameter type_names"
+                ) from e
+
+        # record the current entry index for elements
+        # pointers[i] is the index of entry at (i)th sorted entry lists
+        pointers = [0] * n
+
+        # compare tuple (begin, end, order of type name in input argument
+        # type_names)
+        # we initialize a MinHeap with the first entry of all sorted entry lists
+        # in self.__elements
+        # the metric of comparing entry order is represented by the tuple
+        # (begin index of entry, end index of entry,
+        # the index of the entry type name in input parameter ``type_names``)
+        h: List[Tuple[Tuple[int, int, int], str]] = []
+        for p_idx in range(n):
+            entry_tuple = (
+                (
+                    first_entries[p_idx][constants.BEGIN_INDEX],
+                    first_entries[p_idx][constants.END_INDEX],
+                    p_idx,
+                ),
+                first_entries[p_idx][constants.ENTRY_TYPE_INDEX],
```

Comment:
Is the sorting here performed in the following order:
1. compare begin
2. compare end
3. compare `p_idx`
4. compare name

In docstring, it says the order is:
```
            - start index of the entry.
            - end index of the entry.
            - the index of the entry type name in input parameter ``type_names``.
```

Notice `p_idx` will be different for different types, so the sorting essentially prefers pointers that are smaller, right? This is kinda a weird behavior.

Reply:
`p_idx` is actually the index of the type name in input parameter `type_names`.
Suppose `type_names = ["Token", "Document"]` then the `p_idx` for `Token ` is 0 and `p_idx` for `Document ` is 1.

Reply:
i see, so you put it there for sorting purposes.